src/groupeddataframe/groupeddataframe.jl

```diff
@@ -568,3 +568,177 @@ function Base.get(gd::GroupedDataFrame, key::Union{Tuple, NamedTuple}, default)
+is passed.
+
+See also: [`filter!`](@ref)
+
+# Examples
+```
+julia> df = DataFrame(g=1:2, x='a':'b');
+
+julia> gd = groupby(df, :g)
+GroupedDataFrame with 2 groups based on key: g
+First Group (1 row): g = 1
+│ Row │ g     │ x    │
+│     │ Int64 │ Char │
+├─────┼───────┼──────┤
+│ 1   │ 1     │ 'a'  │
+⋮
+Last Group (1 row): g = 2
+│ Row │ g     │ x    │
+│     │ Int64 │ Char │
+├─────┼───────┼──────┤
+│ 1   │ 2     │ 'b'  │
+
+julia> filter(x -> x.x[1] == 'a', gd)
+GroupedDataFrame with 1 group based on key: g
+First Group (1 row): g = 1
+│ Row │ g     │ x    │
+│     │ Int64 │ Char │
+├─────┼───────┼──────┤
+│ 1   │ 1     │ 'a'  │
+
+julia> filter(:x => x -> x[1] == 'a', gd)
+GroupedDataFrame with 1 group based on key: g
+First Group (1 row): g = 1
+│ Row │ g     │ x    │
+│     │ Int64 │ Char │
+├─────┼───────┼──────┤
+│ 1   │ 1     │ 'a'  │
+
+```
+"""
+Base.filter(f::Base.Callable, gdf::GroupedDataFrame) =
+    gdf[[f(sdf)::Bool for sdf in gdf]]
+Base.filter((col, f)::Pair{<:ColumnIndex}, gdf::GroupedDataFrame) =
+    _filter_helper(gdf, f, gdf.idx, gdf.starts, gdf.ends, parent(gdf)[!, col])
+Base.filter((cols, f)::Pair{<:AbstractVector{Symbol}}, gdf::GroupedDataFrame) =
+    filter([index(parent(gdf))[col] for col in cols] => f, gdf)
+Base.filter((cols, f)::Pair{<:AbstractVector{<:AbstractString}}, gdf::GroupedDataFrame) =
+    filter([index(parent(gdf))[col] for col in cols] => f, gdf)
+Base.filter((cols, f)::Pair, gdf::GroupedDataFrame) =
+    filter(index(parent(gdf))[cols] => f, gdf)
+Base.filter((cols, f)::Pair{<:AbstractVector{Int}}, gdf::GroupedDataFrame) =
+    _filter_helper(gdf, f, gdf.idx, gdf.starts, gdf.ends, (parent(gdf)[!, i] for i in cols)...)
+
+function _filter_helper(gdf::GroupedDataFrame, f, idx::Vector{Int},
+                        starts::Vector{Int}, ends::Vector{Int}, cols...)
+    function mapper(i)
+        idxs = idx[starts[i]:ends[i]]
+        return map(x -> view(x, idxs), cols)
+    end
+
+    if length(cols) == 0
+        throw(ArgumentError("At least one column must be passed to filter on"))
+    end
+    sel = [f(mapper(i)...)::Bool for i in 1:length(gdf)]
+    return gdf[sel]
+end
+
+function Base.filter((cols, f)::Pair{<:AsTable}, gdf::GroupedDataFrame)
+    df_tmp = select(parent(gdf), cols.cols, copycols=false)
+    if ncol(df_tmp) == 0
+        throw(ArgumentError("At least one column must be passed to filter on"))
+    end
+    return _filter_helper_astable(gdf, Tables.columntable(df_tmp), f,
+                                      gdf.idx, gdf.starts, gdf.ends)
+end
+
+function _filter_helper_astable(gdf::GroupedDataFrame, nt::NamedTuple, f,
+                                idx::Vector{Int}, starts::Vector{Int}, ends::Vector{Int})
+    function mapper(i)
+        idxs = idx[starts[i]:ends[i]]
+        return map(x -> view(x, idxs), nt)
+    end
+
+    return gdf[[f(mapper(i))::Bool for i in 1:length(gdf)]]
+end
+
+"""
+    filter!(function, gdf::GroupedDataFrame)
```

Comment:
Do we have any other in-place functions on `GroupedDataFrame` already, apart from `select!` and `transform!`? I'm asking because for consistency with the two latter, it would make sense for `filter!` to drop rows from `parent(gdf)`. Though that could be a unexpected. We need a general policy on this, which could also apply to in-place oeprations on `SubDataFrame`.

Reply:
This is a good question (actually I initially considered to add `filter` only for these reasons). It applies both to `filter` and `filter!`. I have not thought about it because what I proposed was most natural.

So the considerations are:
1. we do not have other functions that mutate `GroupedDataFrame` in-place
2. we do not have functions that mutate `SubDataFrame` schema in-place (of course we allow mutating the data, but not schema)
3. `select!` and `transform!` work only for some `GroupedDataFrames` (ones that have no groups dropped) and they guarantee to retain row count in the parent; also it is in general expected (and documented) that they mutate the parent as they define new columns in general
4. `filter!` and `filter` can work on an already "subsetted" `GroupDataFrame` so then it would be confusing what should happen with groups that are already not present.
5. it is natural to expect that `filter(predicate, collecion)` does the same as `collection[predicate.(collection)]`, which does not mutate the parent of `collection` if it is present.

All in all I think it is better to keep what we have.